specs/simple-serialize.md

```diff
@@ -136,23 +161,47 @@ return serialized_type_index + serialized_bytes
 
 Because serialization is an injective function (i.e. two distinct objects of the same type will serialize to different values) any bytestring has at most one object it could deserialize to. Efficient algorithms for computing this object can be found in [the implementations](#implementations).
 
+Note that deserialization requires hardening against invalid inputs. A non-exhaustive list:
+- Offsets: out of order, out of range, mismatching minimum element size
+- Scope: Extra unused bytes, not aligned with element size.
+- More elements than a list limit allows. Part of enforcing consensus.
+
 ## Merkleization
 
 We first define helper functions:
 
 * `pack`: Given ordered objects of the same basic type, serialize them, pack them into `BYTES_PER_CHUNK`-byte chunks, right-pad the last chunk with zero bytes, and return the chunks.
-* `merkleize`: Given ordered `BYTES_PER_CHUNK`-byte chunks, if necessary append zero chunks so that the number of chunks is a power of two, Merkleize the chunks, and return the root. Note that `merkleize` on a single chunk is simply that chunk, i.e. the identity when the number of chunks is one.
+* `merkleize(data, pad_to)`: Given ordered `BYTES_PER_CHUNK`-byte chunks, if necessary append zero chunks so that the number of chunks is a power of two, Merkleize the chunks, and return the root.
+  There merkleization depends on the input length:
+    - `0` chunks: A chunk filled with zeroes.
+    - `1` chunk: A single chunk is simply that chunk, i.e. the identity when the number of chunks is one.
+    - `1+` chunks: pad to the next power of 2, merkleize as binary tree.
+    - with `pad_to` set: pad the `data` with zeroed chunks to this power of two (virtually for memory efficiency).
```

Comment:
What if `pad_to` is set to a non-power of two (like, `3`)? Are we insisting that max length must be a power of two? Or should `pad_to` be rounded up with the above rules?



E.g. consider a List[uint128, ]

Reply:
rounded up for padding. But limit should be strictly enforced. We really only have power of 2 limits currently, but can imagine it's good to define behavior for non-power of 2 limits.

Reply:
updated it